service/history/workflow/mutable_state_impl.go

```diff
@@ -6726,3 +6729,129 @@ func (ms *MutableStateImpl) applyTombstones(tombstoneBatches []*persistencespb.S
 func (ms *MutableStateImpl) disablingTransitionHistory() bool {
 	return ms.versionedTransitionInDB != nil && len(ms.executionInfo.TransitionHistory) == 0
 }
+
+// GetCurrentDeployment returns the current effective deployment in the following order:
+// RedirectingDeployment takes precedence over DeploymentOverride, over Deployment.
+func (ms *MutableStateImpl) GetCurrentDeployment() *commonpb.WorkerDeployment {
+	versioningInfo := ms.GetExecutionInfo().GetVersioningInfo()
+	if versioningInfo == nil {
+		return nil
+	} else if redirectInfo := versioningInfo.GetRedirectInfo(); redirectInfo != nil {
+		return redirectInfo.GetDeployment()
+	} else if override := versioningInfo.GetDeploymentOverride(); override != nil {
+		return override
+	} else {
+		return versioningInfo.GetDeployment()
+	}
+}
+
+func (ms *MutableStateImpl) GetRedirectInfo() *persistencespb.WorkflowExecutionInfo_VersioningInfo_RedirectInfo {
+	return ms.GetExecutionInfo().GetVersioningInfo().GetRedirectInfo()
+}
+
+func (ms *MutableStateImpl) GetVersioningBehavior() enumspb.VersioningBehavior {
+	versioningInfo := ms.GetExecutionInfo().GetVersioningInfo()
+	if versioningInfo == nil {
+		return enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED
+	} else if redirectInfo := versioningInfo.GetRedirectInfo(); redirectInfo != nil &&
+		redirectInfo.GetBehaviorOverride() != enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED {
+		return redirectInfo.GetBehaviorOverride()
+	} else if versioningInfo.GetBehaviorOverride() != enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED {
+		return versioningInfo.GetBehaviorOverride()
+	} else {
+		return versioningInfo.GetBehavior()
+	}
+}
+
+// StartDeploymentRedirect starts a redirect to the given deployment. If a redirect is already
+// happening, it may be replaced by the new redirect depending on which redirect was made byt a
+// user override. Return's true if the requested redirect is started.
+// TODO: validate source deployment
+func (ms *MutableStateImpl) StartDeploymentRedirect(
+	deployment *commonpb.WorkerDeployment,
+	behaviorOverride enumspb.VersioningBehavior,
+) bool {
+	if deployment.Equal(ms.GetCurrentDeployment()) {
+		// Not a deployment change.
+		return false
+	}
+
+	if redirectInfo := ms.GetRedirectInfo(); redirectInfo != nil {
+		if behaviorOverride == enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED &&
+			redirectInfo.GetBehaviorOverride() != enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED {
+			return false
+		}
+	}
+
+	versioningInfo := ms.GetExecutionInfo().GetVersioningInfo()
+	if versioningInfo == nil {
+		versioningInfo = &persistencespb.WorkflowExecutionInfo_VersioningInfo{}
+		ms.GetExecutionInfo().VersioningInfo = versioningInfo
+	}
+
```

Comment:
Good point, with an ajdustment: when user set's an unpinned override, the next wft might come with a redirect from matching. we should allow that because the wf is unpinned. So the check should be about if the wf is pinned or not rather than the behavior was set by an override or not:
```
	wfBehavior := ms.GetVersioningBehavior()
	if wfBehavior == enumspb.VERSIONING_BEHAVIOR_PINNED &&
		behaviorOverride == enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED {
		// WF is pinned and the redirect is not from a manual override, so we reject it.
		// It's possible that a backlogged task in matching from an earlier time that this wf was
		// unpinned is being dispatched now and wants to redirect the wf. Such task should be dropped.
		return false
	}
```
This also means that the override replacement check I had is invalid so I removed it.